action/protocol/staking/protocol.go

```diff
@@ -96,8 +100,8 @@ func NewProtocol(depositGas DepositGas, sr protocol.StateReader, cfg genesis.Sta
 	}
 
 	return &Protocol{
-		addr:            addr,
-		inMemCandidates: NewCandidateCenter(),
+		addr:        addr,
+		candCenters: cache.NewThreadSafeLruCache(8),
```

Comment:
we should have lru for historical height, and a view for current height which should always be kept in memory.

Reply:
not sure if i understand your comment correctly, but this is the LRU for heights, the latest 8 heights will be kept in this lru cache

view for current height is always in the LRU, so i think not necessary to keep a separate one in memory

for example, if you call ActiveCandidates(ctx context.Context, height uint64) with tip height, getCandCenter() will hit and pull up the correct view for you

the same for ReadState(sr) if sr.Height() = tip height